lib/ffccmmx/client.rb

```diff
@@ -0,0 +1,127 @@
+  class Client
+    V1_ENDPOINT_PREFIX = "/v1/projects/"
+    V1_ENDPOINT_SUFFIX = "/messages:send"
+    TOPIC_ENDPOINT_PREFIX = "/iid/v1"
+    TOPIC_BATCH_SUBSCRIBE_SUFFIX = ":batchAdd"
+    TOPIC_BATCH_UNSUBSCRIBE_SUFFIX = ":batchRemove"
+
+    attr_reader :base_url, :topic_base_url, :push_uri, :configuration, :access_token, :access_token_expiry, :httpx
+
+    # rubocop:disable Metrics/MethodLength
+    def initialize(base_url, topic_base_url, project_id, configuration) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+      @base_url = base_url
+      @topic_base_url = topic_base_url
+      @project_id = project_id
+
+      @push_uri = URI.join(base_url, V1_ENDPOINT_PREFIX + project_id.to_s + V1_ENDPOINT_SUFFIX)
+      @batch_subscribe_uri = create_batch_uri(:subscribe)
+      @batch_unsubscribe_uri = create_batch_uri(:unsubscribe)
+
+      @configuration = configuration.dup
+      access_token_response = v1_authorize
+      @access_token = access_token_response["access_token"]
+      @access_token_expiry = Time.now.utc + access_token_response["expires_in"]
+      @httpx = ::HTTPX.plugin(:auth).plugin(:persistent, configuration.httpx_options)
+    end
+    # rubocop:enable Metrics/MethodLength
+
+    def push(body, query: {}, headers: {})
+      do_push_request(body, query, headers)
+    end
+
+    def subscribe(topic, *instance_ids, query: {}, headers: {})
+      do_subscription_request(topic, *instance_ids, :subscribe, query, headers)
+    end
+
+    def unsubscribe(topic, *instance_ids, query: {}, headers: {})
+      do_subscription_request(topic, *instance_ids, :unsubscribe, query, headers)
+    end
+
+    private
+
+    def create_batch_uri(action)
+      case action
+      when :subscribe
+        URI.join(topic_base_url, TOPIC_ENDPOINT_PREFIX + TOPIC_BATCH_SUBSCRIBE_SUFFIX)
+      when :unsubscribe
+        URI.join(topic_base_url, TOPIC_ENDPOINT_PREFIX + TOPIC_BATCH_UNSUBSCRIBE_SUFFIX)
+      else
+        raise ArgumentError, "Invalid action: #{action}. Use :subscribe or :unsubscribe."
+      end
+    end
+
+    def v1_authorize
+      @auth ||= if configuration.json_key_io
+                  Google::Auth::ServiceAccountCredentials.make_creds(
+                    json_key_io: prepare_json_key_io,
+                    scope: configuration.scope
+                  )
+                else
+                  # from ENV
+                  Google::Auth::ServiceAccountCredentials.make_creds(scope: configuration.scope)
+                end
+      @auth.fetch_access_token
+    end
+
+    def prepare_json_key_io
+      io = if configuration.json_key_io.respond_to?(:read)
+             configuration.json_key_io
+           else
+             File.open(configuration.json_key_io)
+           end
+      io.rewind if io.respond_to?(:read)
+      io
+    end
+
+    def do_push_request(json, query, headers)
+      access_token_refresh
+      uri = @push_uri.dup
+      uri.query = URI.encode_www_form(query) unless query.nil? || query.empty?
```

Comment:
呼び出し元の型的にはnil来ることなさそう

```suggestion
      uri =
        if query.empty?
          @push_uri
        else
          uri = @push_uri.dup
          uri.query = URI.encode_www_form(query)
        end
```

Reply:
そもそもFCMのpushのエンドポイント呼び出すときにqueryって設定することあるんだっけ